enterprise/server/usage/usage.go

```diff
@@ -0,0 +1,138 @@
+package usage
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/tables"
+	"github.com/buildbuddy-io/buildbuddy/server/util/perms"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/buildbuddy-io/buildbuddy/server/util/timeutil"
+	"github.com/go-redis/redis"
+)
+
+const (
+	// collectionPeriodDuration determines the granularity at which we flush data
+	// to the DB.
+	//
+	// NOTE: if this value is changed, then the implementation of
+	// collectionPeriodStartUsec needs to change as well. Also note that if this
+	// value is increased, we may lose some usage data for a short period of time,
+	// since keys at smaller granularity may not be properly flushed to the DB.
+	collectionPeriodDuration = 1 * time.Minute
+
+	// redisKeyTTL defines how long usage keys have to live before they are
+	// deleted automatically by Redis.
+	//
+	// Keys should live for at least 2 collection periods since collection periods
+	// aren't finalized until the period is past, plus some wiggle room for Redis
+	// latency. We add a few more collection periods on top of that, in case
+	// flushing fails due to transient errors.
+	redisKeyTTL = 5 * collectionPeriodDuration
+
+	redisUsageKeyPrefix  = "usage/"
+	redisGroupsKeyPrefix = redisUsageKeyPrefix + "groups/"
+	redisCountsKeyPrefix = redisUsageKeyPrefix + "counts/"
+
+	redisTimeKeyFormat = time.RFC3339
+)
+
+type TrackerOpts struct {
+	Clock timeutil.Clock
+}
+
+type tracker struct {
+	env   environment.Env
+	rdb   *redis.Client
+	clock timeutil.Clock
+}
+
+func NewTracker(env environment.Env, opts *TrackerOpts) (*tracker, error) {
+	rdb := env.GetCacheRedisClient()
+	if rdb == nil {
+		return nil, status.UnimplementedError("Missing redis client for usage")
+	}
+	clock := opts.Clock
+	if clock == nil {
+		clock = timeutil.NewClock()
+	}
+	return &tracker{
+		env:   env,
+		rdb:   rdb,
+		clock: clock,
+	}, nil
+}
+
+func (ut *tracker) Increment(ctx context.Context, counts *tables.UsageCounts) error {
+	groupID, err := perms.AuthenticatedGroupID(ctx, ut.env)
+	if err != nil {
+		if perms.IsAnonymousUserError(err) && ut.env.GetConfigurator().GetAnonymousUsageEnabled() {
+			// Don't track anonymous usage for now.
+			return nil
+		}
+		return err
+	}
+
+	m, err := countsToMap(counts)
+	if err != nil {
+		return err
+	}
+	if len(m) == 0 {
+		return nil
+	}
+
+	t := ut.clock.Now()
+
+	pipe := ut.rdb.TxPipeline()
+	// Add the group ID to the set of groups with usage
+	groupsKey := groupsRedisKey(t)
+	pipe.SAdd(ctx, groupsKey, groupID)
+	pipe.Expire(ctx, groupsKey, redisKeyTTL)
+	// Increment the hash values
+	countsKey := countsRedisKey(groupID, t)
+	for k, c := range m {
```

Comment:
nit: this could be made more readable by naming your loop vars and map. For example:

for countField, count := range counts {
  pipe.HIncrBy(ctx, groupIDCollectionPeriod, countField, count)
}

Reply:
Done